examples/sparse_gamma_def.py

```diff
@@ -0,0 +1,153 @@
+import torch
+
+import pyro
+import pyro.optim as optim
+import pyro.poutine as poutine
+import wget
+from pyro.distributions import Gamma, Poisson
+from pyro.infer import SVI, Trace_ELBO
+
+torch.set_default_tensor_type('torch.FloatTensor')
+pyro.enable_validation(True)
+pyro.util.set_rng_seed(0)
+
+
+class SparseGammaDEF(object):
+    def __init__(self):
+        # define the sizes of the layers in the deep exponential family
+        self.top_width = 100
+        self.mid_width = 40
+        self.bottom_width = 15
+        self.image_size = 64 * 64
+        # define hyparameters that control the prior
+        self.alpha_z = torch.tensor(0.1)
+        self.beta_z = torch.tensor(0.1)
+        self.alpha_w = torch.tensor(0.1)
+        self.beta_w = torch.tensor(0.3)
+        # define parameters used to initialize variational parameters
+        self.alpha_init = 0.5
+        self.mean_init = 0.0
+        self.sigma_init = 0.1
+        self.softplus = torch.nn.Softplus()
+
+    # define the model
+    @poutine.broadcast
+    def model(self, x):
+        x_size = x.size(0)
+
+        # sample the global weights
+        with pyro.iarange("w_top_iarange", self.top_width * self.mid_width):
+            w_top = pyro.sample("w_top", Gamma(self.alpha_w, self.beta_w))
+        with pyro.iarange("w_mid_iarange", self.mid_width * self.bottom_width):
+            w_mid = pyro.sample("w_mid", Gamma(self.alpha_w, self.beta_w))
+        with pyro.iarange("w_bottom_iarange", self.bottom_width * self.image_size):
+            w_bottom = pyro.sample("w_bottom", Gamma(self.alpha_w, self.beta_w))
+
+        # sample the local latent random variables
+        # (the iarange encodes the fact that the z's for different datapoints are conditionally independent)
+        with pyro.iarange("data", x_size):
+            z_top = pyro.sample("z_top", Gamma(self.alpha_z, self.beta_z).expand([self.top_width]).independent(1))
+            mean_mid = torch.mm(z_top, w_top.reshape(self.top_width, self.mid_width))
+            z_mid = pyro.sample("z_mid", Gamma(self.alpha_z, self.beta_z / mean_mid).independent(1))
+            mean_bottom = torch.mm(z_mid, w_mid.view(self.mid_width, self.bottom_width))
+            z_bottom = pyro.sample("z_bottom", Gamma(self.alpha_z, self.beta_z / mean_bottom).independent(1))
+            mean_obs = torch.mm(z_bottom, w_bottom.view(self.bottom_width, self.image_size))
+
+            # observe the data using a poisson likelihood
+            pyro.sample('obs', Poisson(mean_obs).independent(1), obs=x)
+
+    # define the guide a.k.a. variational distribution.
+    # (note the guide is mean field)
+    @poutine.broadcast
+    def guide(self, x):
+        x_size = x.size(0)
+
+        # define a helper function to sample z's for a single layer
+        def sample_zs(name, width):
+            alpha_z_q = pyro.param("log_alpha_z_q_%s" % name,
+                                   torch.tensor(self.alpha_init * torch.ones(x_size, width) +
```

Comment:
nit: What is the purpose of the `torch.tensor(...)` wrapper here and in three places below? I'm surprised this is needed.

Also I would recommend wrapping the tensor init arg in a `lambda: ...` to avoid calling `torch.randn` on later iterations.

Reply:
well don't i need it so that the param is a proper leaf node?

Reply:
I believe as of #908 `pyro.param` automatically creates a leaf.

Reply:
ok, done